Improved text location information for UIA, Word and compound text infos

LeonarddeR opened the pull request:
### Link to issue number:
Fixes #7916 
Fixes #8370 
Fixes #8371

### Summary of the issue:
NVDA's location information for text info is primarily based on screen points. For example, offset based text info implementations use the _getPointFromOffset function to get the screen coordinates for a particular point.

By default however, the display model and IAccessible2 offer per character position information based on rectangles instead of points, which is more accurate than points. For example, when creating a virtual caret in Firefox and Chrome virtual buffers, the virtual caret has to be positioned at the right of the currently selected character. Furthermore, based on rectangle information, it is possible to calculate the bounding rectangle of an entire text range. This makes it possible to the get the bounding rectangle of a selection in order for a magnifier to track to, a highlight to be positioned correctly, a virtual selection to be shown visually, etc.

Non-offset based text controls, like UIA based controls and MS Word, do expose per text range bounding rectangle information. For Word, you can retrieve the bounding rectangle of a text range, whereas for UIA, there are per line bounding rectangles within the text range. These cases are currently not implemented in NVDA, which means that the route mouse to navigator object is only able to route the mouse to the object itself, rather than the exact character the review cursor is at.

### Description of how this pull request fixes the issue:
This pr builds upon locationHelper and is the predecessor of another huge pr that will implement focus and browse mode caret highlighting in the core of NVDA. It will also add a framework that lays the foundation to implement support for third party magnifier software or APIs in the future.

This pr contains the following.

1. Added a boundingRect property to textInfos.

	+ NVDAObjectTextInfo returns the location of the object.

	+ The Word TextInfo object gets the bounding rectangle from the text range.

	+ UIATextInfo gets the per line bounding rectangles from the text range and creates a bounding 
rectangle of the full text range using locationHelper

	+ DisplayModelTextInfo creates a bounding rectangle from the rectangles in the range.

	+ OffsetsTextInfo contains the most complex implementation. It calculates a bounding rectangle of the text range based on the locations of:

		* The start of the range
		* The end of the range
		* The start and end of every line in the range.

	As this is very performance heavy for huge text ranges, I added _getFirstVisibleOffset and _getLastVisibleOffset methods that can help in limiting the range to the characters that are visible on screen. This works properly for EditTextInfo, for IA2TextInfo the support is more limited. However, I tried to make the first and last visible offset methods as conservative as possible, (i.e. when the retrieved offset might not be reliable, a LookupError is raised.

	+ TreeCompoundTextInfo calculates a bounding rectangle based on the bounding rectangles of the underlying text infos.

	+ MozillaCompoundTextInfo tries to do the same, but its implementation differs from TreeCompoundTextInfo.

2. Specific for UIA:

    + It is now again possible to create a TextInfo object from a point. This support was disabled in the code as the client library was said to freeze under Windows 7. I've not been able to reproduce these freezes under Windows 10. I re-enabled this for now, however this requires some testing under Windows 7. @MichaelDCurran pointed out that textarea nodes under Internet Explorer 11 might prove a good test case. @josephsl: WOuld you be able to assist with this?
    + STATE_OFFSCREEN is now properly added to UIA objects that have the OffSccreen property.

3. The base implementation of TextInfo.pointAtStart now tries to get the top left corner of the boundingRect property. Also, for OffsetsTextInfo, pointAtStart first tries to get the upper left corner of the start offset bounding rectangle before using _getOffsetFromPoint.

4. For OffsetsTextInfo, the base implementation of _getPointFromOffset uses the center of the rectangle retrieved with _getBpoundingRectFromOffset. For IA2TextInfo and DisplayModelTextInfo, _getPointFromOffset has been removed and is replaced by an implementation of _getBoundingRectFromOffset.

5. A new hasRelevantLocation property has been added to NVDAObject. This property will be used to determine whether the location of an object can be used by a highlighter or magnifier to highlight or track to an object, respectively. This property returns True if the following conditions are met:

	* The object isn't off screen
	* The object doesn't have the unavailable state
	* The location is not None
	* At least one location coordinate is nonzero

	This property is used by CompoundTextInfos to ignore the bounding rectangle of objects that are off screen or unavailable while calculating the bounding rectangle of a compound text range.

### Testing performed:
1. Tested the boundingRect property and _getBoundingRectFromOffset implementations for the cases described above.
2. IN Edge browse mode and Word, tested the move mouse to navigator object command. The mouse echo corresponded with the character at the caret.

### Known issues with pull request:
1. As noted above, when creating an UIA text info from a point under Windows 7 at least, freezes might occur. This has to be tested before this behavior can be enabled in release versions of NVDA.
2. Especially for compound text, the performance of bounding rectangle retrieval might be slow.

### Change log entry:
* Bug fixes
    + The functionality of the move mouse to navigator object command has been majorly improved in Microsoft Word as well as for UIA controls, particularly Microsoft Edge. (#7916, #8371)
    + The echo of text under the mouse has been improved within Microsoft Edge and other UIA applications. (#8370)

* Changes for developers
    + OffsetsTextInfo objects can now implement the _getBoundingRectFromOffset method to allow retrieval of bounding rectangles per characters instead of points.
    + Added a boundingRect property to TextInfo objects to retrieve the bounding rectangle of a range of text. (#8371)

Commit:
Created bounding rectangle text info property and implemented it for Word, UIA, offsets and Mozilla compound documents

josephsl commented:
Hi,

I'd like to recuse myself from reviewing this, as I'm not quite that versed in text infos and rectangles yet.

Thanks.

leonardder commented:
No problem. Would you be able to test the possible Windows 7 issue that 
might apply?

josephsl commented:
Hi, I need a test case (hopefully the issue number) so I can test it on a VM. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, July 30, 2018 8:03 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Review requested <review_requested@noreply.github.com>
Subject: Re: [nvaccess/nvda] Improved text location information for UIA, Word and compound text infos (#8572)

 

No problem. Would you be able to test the possible Windows 7 issue that 
might apply?



—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8572#issuecomment-408895138> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkC3drNZSyuiKF7UlQbxD3V-o7_zCks5uLyAbgaJpZM4VmgTp> .

leonardder commented:
I'm afraid there's neither an issue nor a valid test case for this, as this is only based on a comment in the code:

> #rangeFromPoint causes a freeze in UIA client library!

Internet Explorer 11 text inffo ffor textarea nodes is based on UIA, so in Windows 7, that'd be te way to test this.

josephsl commented:
Hi,

I see. I guess what we need is steps to reproduce possible issues in Windows 7+IE11+specific website+specific content.

By the way, is the latest master branch merged?

Thanks.

josephsl commented:
Hi,

@MichaelDCurran, do you happen to have the URL for a website where we use UIA for textarea node instead of MSHTML? If so, let us know so we can try the following:

Requiremenets:

* One must be testing babbagecom/boundingRectangle branch (hopefully with latest master commits merged).
* Must be running Windows 7 SP1, Windows 8.1, or Windows 10. Windows 8 does not qualify because support for it has ended in 2016.
* If using Windows 10, preferably all releases (build 10240 onwards, although build 15063 or later is prefered due to support policy from Microsoft).

Steps:

1. Navigate to the said website using Internet Explorer and Microsoft Edge.
2. Locate the textarea.
3. Verify that developer info says text info is a UIA text info instance.
4. Open Python Console and type the following as exactly as shown:

import NVDAObjects.UIA
import textInfos
NVDAObjects.UIA.UIATextInfo(nav, textInfos.Point(randomX, randomY))

Where "randomX" and "randomY" can be any coordinate. Ideally you should get a text info in return, but on my computer (tested with Edge on Windows 10 build 17713) with random coords set to (400, 400), I get:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "NVDAObjects\UIA\__init__.pyc", line 316, in __init__
AttributeError: 'NoneType' object has no attribute 'RangeFromPoint'

Thanks.

josephsl commented:
Hi,

Regarding merging master commits: I see that it was rebased on master branch, so ignore my comment.

Thanks.

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [30 Jul 2018, 20:04 CEST](https://github.com/nvaccess/nvda/pull/8572#issuecomment-408956352 "2018-07-30T18:04:46Z - Replied by Github Reply Comments"):
> 3. Verify that developer info says text info is a UIA text info instance.

This should be NVDAObjects.IAccessible.MSHTML.UIAMSHTMLTextInfo

> I get:
> 
> Traceback (most recent call last):
>   File "<console>", line 1, in <module>
>   File "NVDAObjects\UIA\__init__.pyc", line 316, in __init__
> AttributeError: 'NoneType' object has no attribute 'RangeFromPoint'

Wow, let me look into that one, this is certainly not what I expect to happen.

leonardder commented:
This might be an easier test case.

1. In Internet Explorer, open this particular issue and login to github.
2. Focus the comment text area
3. Make sure the objects text info is an UIAMSHTMLTextInfo
4. Focus the multi line edit control in focus mode
5. Make sure mouse tracking is on and resolution is set to character.
6. Type some random text
7. Walk to the text with your right arrow key, and for every character, move the mouse to it with the move mouse to navigator object command (NVDA+numpad divide or laptop NVDA+shift+m). 

If all goes well, NVDA should not freeze and mouse tracking should pronounce the character where the caret is located.

josephsl commented:
Hi,

Yes, no longer freezes in Windows 7+IE11. We still want to make sure older versions such as 9 and 10 are still covered.

Thanks.

leonardder commented:
According to Mick, this uia implementation is only used as part of IE11., if I understood him correctly.

> Op 30 jul. 2018 om 20:59 heeft Joseph Lee <notifications@github.com> het volgende geschreven:
> 
> Hi,
> 
> Yes, no longer freezes in Windows 7+IE11. We still want to make sure older versions such as 9 and 10 are still covered.
> 
> Thanks.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

derekriemer reviewed (commented):
Good job.

Commit:
Updated changes.t2t

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

Commit:
Revert changes.t2t updates

feerrenrut commented:
@leonardder Is the blocked label still appropriate?

I'm holding off on reviewing this, waiting for the comments by @derekriemer to be addressed.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle, fixing merge conflicts

Commit:
Update adobe reader vbuf

leonardder commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -116,17 +116,30 @@ class IA2TextTextInfo(textInfos.offsets.OffsetsTextInfo):
+		return RectLTWH(*res)
 
 	def _get_unit_mouseChunk(self):
 		return "mouseChunk"
```
I have no idea. Note that I did not touch this code.

leonardder commented on source/NVDAObjects/IAccessible/ia2TextMozilla.py:
```diff
@@ -644,3 +645,27 @@ def _get_NVDAObjectAtStart(self):
+				break
+			try:
+				ti, obj = self._findNextContent(ti)
+			except LookupError:
```
I dont'follow how you'd like to see this. Getting the bounding rectangle for a range requires the complete text info to be walked, so a generator wouldn't add any performance improvement.

leonardder commented on source/NVDAObjects/IAccessible/ia2TextMozilla.py:
```diff
@@ -644,3 +645,27 @@ def _get_NVDAObjectAtStart(self):
+				# Can't move forward any further.
+				break
+			if obj == self._endObj:
+				ti = self._end
```
Because that ti corresponds with the end of the range and its endOffset is the end of the current range. The earlier ti that is assigned by the call of _findNextContent is a textInfo for the object, but it might contain text that is out of the current range. I will add a comment to clarify this.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -311,10 +311,9 @@ def __init__(self,obj,position,_rangeObj=None):
-			#p=POINT(position.x,position.y)
-			#self._rangeObj=self.obj.UIATextPattern.RangeFromPoint(p)
-			raise NotImplementedError
+			#rangeFromPoint used to cause a freeze in UIA client library!
```
The code is not new, it is no longer commented out. I'm still not convinced that this bug in the UIA client library is fixed for all cases that we could encounter.

leonardder replied:
Some tests are performed on Windows 7 to see whether these freezes still occur, and they haven't been found reproducible. If we don't reintroduce this, that will instantly break routing the mouse to text positions in Edge again, for example.

We could limit this to a particular version of UIA if desired. Unfortunately there is no reference to a specific list of steps to reproduce this.

leonardder commented on source/displayModel.py:
```diff
@@ -489,6 +487,20 @@ def getTextInChunks(self,unit):
+		return RectLTWH.fromCollection(*rects).toPhysical(self.obj.windowHandle)
+
+	def _getFirstVisibleOffset(self):
+		return 0
```
Because this is the display model, all characters within the model are visible on the screen.

leonardder commented on source/textInfos/offsets.py:
```diff
@@ -497,3 +569,23 @@ def updateSelection(self):
+			obj = obj.rootNVDAObject
+		if obj.hasIrrelevantLocation:
+			raise LookupError("Object is off screen, invisible or has no location")
+		exclusiveX, exclusiveY = obj.location.bottomRight
```
Could you elaborate on your question? I'm afraid I'm not getting it. The bounding rectangle is supposed to only cover the text within the range that's currently visible. If the end offset of the range is lower than the last visible offset, the former is used to calculate the bounding rectangle.

Commit:
Review actions

leonardder commented:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [20 Aug 2018, 05:36 CEST](https://github.com/nvaccess/nvda/pull/8572#issuecomment-414188668 "2018-08-20T03:36:20Z - Replied by Github Reply Comments"):
> @leonardder Is the blocked label still appropriate?

I added the blocked label since we didn't know enough about possible issues with the UIA client library, however testing reveals that it might no longer be an issue for us. I removed it.

> I'm holding off on reviewing this, waiting for the comments by @derekriemer to be addressed.

I think I've been able to do that just yet.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

leonardder commented:
@feerrenrut: I just wanted to point out that the vision framework which I"m currently working on will include a focus highlight feature. It is currently blocked by this pr, so your review would be greatly appreciated. Note that @derekriemer has already performed some review.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

Commit:
Implement bounding rectangles for content recog

Commit:
Update unit tests to work properly

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

Commit:
Add basic locationText

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

Commit:
Base UIA word locationText on pointAtStart property

leonardder commented:
@michaeldcurran: As this is a quite big change in terms of code additions, I'm also requesting review from you, since you know the textInfos code pretty well.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle, fixing merge conflicts

leonardder commented:
I think your point makes sense. The current code will indeed create a 
bounding rect that may contain coordinates that fall outside the range. 
I now see that having several bounding rectangles is also beneficial for 
magnifiers when it comes to text selection (i.e. based on is selection 
anchored at start).

Would you prefer a boundingRects property that behaves similar to the 
UIA boundingRects property?

michaelDCurran commented:
Similar, though I think UIA's boundingRects property is flattened to one 
1d array. We can obviously do better than that and have a list of rects.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

Commit:
Rename boundingRect property to boundingRects

Commit:
MozillaCompoundTextInfo: calculate boundingRects with a copy, since _findNextContent manipulates the textInfo instance

Commit:
Since word doesn't support per line bounding rectangles, just stick to a pointAtStart implementation

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [2 Nov 2018, 00:27 CET](https://github.com/nvaccess/nvda/pull/8572#issuecomment-435220581 "2018-11-01T23:27:34Z - Replied by Github Reply Comments"):
> I have not read through the code too carefully yet, but I am concerned about  having a boundingRect property on a textInfo rather than a boundingRects property. If you have a range that starts say  a few characters in from the end of one line, and extends a few characters onto the second line, then the boundingRect would include the entirity of both lines. Is This really what you want for highlighting / magnification? I would have thought you'd want two rects, one for the end of the first line, and one for the beginning of the second? Most APIs I have seen  expose boundingRects not a boundingRect for this reason.

@michaeldcurran: This is now addressed, so if you could have a look at this again, that'd be great!

feerrenrut reviewed (commented):
Have you tested this with different screen scaling in Windows?

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

leonardder commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -118,21 +118,34 @@ class IA2TextTextInfo(textInfos.offsets.OffsetsTextInfo):
+			return offset
 		else:
-			raise NotImplementedError
+			raise LookupError
```
It is actually exactly the other way around. offsetFromPoint is expected to raise LookupErrors, see for example event_mouseMove in the base NVDAObject, which is responsible for speaking text under the mouse. In the case where a RuntimeError was raised, this error wasn't caught and would result into a massive amount of errors whenever you move the mouse within such an object. So I believe that RuntimeError was wrong here.

leonardder commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -638,6 +637,29 @@ def getTextWithFields(self,formatConfig=None):
+		if not rectArray:
+			return rects
+		rects.extend(
+			RectLTWH.fromFloatCollection(*rectArray[i:i+4])
```
I will address this, with the exception of using a rect generator instead of a list.

leonardder commented on source/NVDAObjects/window/edit.py:
```diff
@@ -174,30 +174,31 @@ def _getPointFromOffset(self,offset):
-		(left,top,width,height)=self.obj.location
-		point.x=point.x+left
-		point.y=point.y+top
+		if -1 in (point.x, point.y):
```
See the remarks section [here](https://docs.microsoft.com/en-us/windows/desktop/controls/em-posfromchar). That actually made me change the check to point.x<0 or point.y<0. Thanks for catching this!

leonardder commented on source/NVDAObjects/window/winword.py:
```diff
@@ -1000,6 +1001,19 @@ def move(self,unit,direction,endPoint=None):
+		left = ctypes.c_int()
+		top = ctypes.c_int()
+		width = ctypes.c_int()
+		height = ctypes.c_int32()
```
Ugh, it doesn't matter at all since they evaluate both to c_long, but I fixed this, thanks!

Commit:
Review actions

leonardder commented:
@feerrenrut: re your question about different display scaling, I haven't yet tested this. I assume I should do this for Word and UIA. For IA2, we already had code for pointFromOffset

feerrenrut reviewed (approved):
> re your question about different display scaling, I haven't yet tested this. I assume I should do this for Word and UIA. For IA2, we already had code for pointFromOffset

Yes, that sound fair. If you believe the testing to be adequate here, then I'm happy for this to be merged.

leonardder commented:
Thanks for approving!


I tested in Word with and without UIA by moving the mouse to several 
characters in a sentence. My colleague verified visually that the mouse 
ended up at the right character. Also mouse echo reported the right 
character.

I think and hope we're good to go.

Commit:
Merge remote-tracking branch 'origin/master' into boundingRectangle

feerrenrut committed:
Update changes file for PR #8572

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into PR8572
